### PR TITLE
adds module for dleq proofs

### PR DESCRIPTION
**Pull request by armfazh**
Add DLEQ proofs of knowledge.

### CODE REVIEW

**Review by lukevalenta — APPROVED** [review]
I confirmed that the implementation matches https://www.ietf.org/id/draft-irtf-cfrg-voprf-09.html#name-discrete-log-equivalence-pr (and also read https://blog.cloudflare.com/privacy-pass-the-math/ for some additional context).

The code itself LGTM, but there's no documentation. Adding some comments describing how the functions translate to the draft specification would be helpful. I'll leave some comments inline with more detailed suggestions.

File: src/dleq.ts [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+    readonly dst: string
+}
+
+const LABELS = {
+    Seed: 'Seed-',
+    Challenge: 'Challenge',
+    Composite: 'Composite',
+    HashToScalar: 'HashToScalar-'
+} as const
+
```
Comment (lukevalenta) [inline src/dleq.ts:21]:
```suggestion
// computeComposites implements ComputeComposites and ComputeCompositiesFast functions from https://www.ietf.org/id/draft-irtf-cfrg-voprf-09.html#name-discrete-log-equivalence-pr.
```

File: src/dleq.ts [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+        }
+    }
+
+    if (key) {
+        Z = Group.mul(key, M)
+    }
+
+    return { M, Z }
+}
+
```
Comment (lukevalenta) [inline src/dleq.ts:67]:
```suggestion
// challenge implements the shared subprocedure for generating a challenge used by the GenerateProof and VerifyProof functions from https://www.ietf.org/id/draft-irtf-cfrg-voprf-09.html#name-discrete-log-equivalence-pr to generate a challenge from the input elements. The point arguments correspond to [B, M, Z, t2, t3] from the specification.
```

File: src/dleq.ts [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    constructor(
+        public readonly params: DLEQParams,
+        public readonly c: Scalar,
+        public readonly s: Scalar
+    ) {}
+
+    verify(p0: [Elt, Elt], p1: [Elt, Elt]): Promise<boolean> {
+        return this.verify_batch(p0, [p1])
+    }
+
```
Comment (lukevalenta) [inline src/dleq.ts:98]:
```suggestion
// verify_batch implements the VerifyProof function from https://www.ietf.org/id/draft-irtf-cfrg-voprf-09.html#name-discrete-log-equivalence-pr. The argument p0 corresponds to the elements A, B, and the argument p1s corresponds to the arrays of elements C and D from the specification.
```

File: src/dleq.ts [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    }
+}
+
+export class DLEQProver {
+    constructor(public readonly params: DLEQParams) {}
+
+    prove(k: Scalar, p0: [Elt, Elt], p1: [Elt, Elt], r?: Scalar): Promise<DLEQProof> {
+        return this.prove_batch(k, p0, [p1], r)
+    }
+
```
Comment (lukevalenta) [inline src/dleq.ts:114]:
```suggestion
// prove_batch implements the GenerateProof function from https://www.ietf.org/id/draft-irtf-cfrg-voprf-09.html#name-discrete-log-equivalence-pr. The argument p0 corresponds to the elements A, B, and the argument p1s corresponds to the arrays of elements C and D from the specification.
```

File: src/dleq.ts [hunk 5]
```diff
@@ -0,0 +1,129 @@
+// Copyright (c) 2021 Cloudflare, Inc. and contributors.
+// Copyright (c) 2021 Cloudflare, Inc.
+// Licensed under the BSD-3-Clause license found in the LICENSE file or
+// at https://opensource.org/licenses/BSD-3-Clause
+
```
Comment (lukevalenta) [inline src/dleq.ts:5]:
```suggestion
//
// Implementation of batched discrete log equivalents proofs (DLEQ) as described in https://www.ietf.org/id/draft-irtf-cfrg-voprf-09.html#name-discrete-log-equivalence-pr.
```

**Review by lukevalenta — APPROVED** [review]

**Review by cdrubin — COMMENTED** [review]
Please add a link to the paper and page or similar where many of these symbols names come from so that variables like M, Z and Ci, Di have context.

**Comment by armfazh** [comment]
> Please add a link to the paper and page or similar where many of these symbols names come from so that variables like M, Z and Ci, Di have context.

This was covered by documentation comments added by Luke pointing to the draft's URL.

**Pull request merged**